File: algorithm/力扣热题100/3_无重复字符的最长子串.js

var lengthOfLongestSubstring = function (s) {
    const win = new Map()
    let res = 0
    let i = 0, j = 0;
    while (j < s.length) {
        const ch = s[j]
        win.set(ch, (win.get(ch) || 0) + 1)
        j++

        while (win.get(ch) > 1) {
            win.set(s[i], win.get(s[i]) - 1)
            i++;
        }
        res = Math.max(res, j - i)
    }
    return res
}

File: algorithm/力扣热题100/3_无重复字符的最长子串.go
package main

func lengthOfLongestSubstring(s string) int {
	win := map[byte]int{}
	i := 0
	j := 0
	res := 0
	for j < len(s) {
		ch := s[j]
		win[ch]++
		j++

		for i < len(s) && win[s[j-1]] > 1 {
			win[s[i]]--
			i++
		}

		if j-i > res {
			res = j - i
		}
	}
	return res
}

// func main() {
// 	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
// }
